swarm/src/connection.rs

```diff
@@ -186,10 +189,22 @@ where
         max_negotiating_inbound_streams: usize,
         idle_timeout: Duration,
     ) -> Self {
-        let initial_protocols = gather_supported_protocols(&handler);
-        if !initial_protocols.is_empty() {
+        let local_supported_protocols = handler
+            .listen_protocol()
+            .upgrade()
+            .protocol_info()
+            .map(|i| (AsStrHashEq(i), true))
+            .collect::<HashMap<_, _>>();
+
+        if !local_supported_protocols.is_empty() {
+            let temp = local_supported_protocols
+                .keys()
+                .filter_map(|i| StreamProtocol::try_from_owned(i.0.as_ref().to_owned()).ok())
+                .collect::<Vec<_>>();
             handler.on_connection_event(ConnectionEvent::LocalProtocolsChange(
-                ProtocolsChange::Added(ProtocolsAdded::from_set(&initial_protocols)),
```

Comment:
Can we re-introduce an API on `ProtocolsAdded` to:

a) make the diff smaller
b) remove this low-level noise from this `poll` function

This function is complicated enough as it is. Anything we can do to hide implementation details is good. I'd suggest using the previous approach of passing `initial_protocols` to a constructor function.

Reply:
Hope its better now

Reply:
Thanks!